server/src/main/java/org/elasticsearch/index/fielddata/ordinals/GlobalOrdinalsIndexFieldData.java

```diff
@@ -22,53 +22,64 @@
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.OrdinalMap;
 import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.TermsEnum;
 import org.apache.lucene.search.SortField;
 import org.apache.lucene.util.Accountable;
 import org.elasticsearch.common.Nullable;
 import org.elasticsearch.index.AbstractIndexComponent;
 import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.fielddata.AtomicOrdinalsFieldData;
-import org.elasticsearch.index.fielddata.IndexFieldData;
 import org.elasticsearch.index.fielddata.IndexFieldData.XFieldComparatorSource.Nested;
 import org.elasticsearch.index.fielddata.IndexOrdinalsFieldData;
 import org.elasticsearch.index.fielddata.ScriptDocValues;
 import org.elasticsearch.index.fielddata.plain.AbstractAtomicOrdinalsFieldData;
 import org.elasticsearch.search.MultiValueMode;
 
+import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.function.Function;
 
 /**
- * {@link IndexFieldData} base class for concrete global ordinals implementations.
+ * Concrete implementation of {@link IndexOrdinalsFieldData} for global ordinals.
+ * A single instance of this class should be used to cache global ordinals per {@link DirectoryReader}.
+ * However {@link #loadGlobal(DirectoryReader)} always creates a new instance of {@link Consumer} from the cached
+ * value in order to reuse the segment's {@link TermsEnum} that are needed to retrieve terms from global ordinals.
+ * Each instance of {@link Consumer} uses a new set of {@link TermsEnum} that can be reused during the collection,
+ * this is done to avoid creating all segment's {@link TermsEnum} each time we want to access the values of a single
+ * segment.
  */
-public class GlobalOrdinalsIndexFieldData extends AbstractIndexComponent implements IndexOrdinalsFieldData, Accountable {
+public final class GlobalOrdinalsIndexFieldData extends AbstractIndexComponent implements IndexOrdinalsFieldData, Accountable {
 
     private final String fieldName;
     private final long memorySizeInBytes;
 
     private final OrdinalMap ordinalMap;
-    private final Atomic[] atomicReaders;
+    private final AtomicOrdinalsFieldData[] segmentAfd;
     private final Function<SortedSetDocValues, ScriptDocValues<?>> scriptFunction;
 
-
-    protected GlobalOrdinalsIndexFieldData(IndexSettings indexSettings, String fieldName, AtomicOrdinalsFieldData[] segmentAfd,
-                                           OrdinalMap ordinalMap, long memorySizeInBytes, Function<SortedSetDocValues,
-                                           ScriptDocValues<?>> scriptFunction) {
+    protected GlobalOrdinalsIndexFieldData(IndexSettings indexSettings,
```

Comment:
Small comment, I think this can be package-private now that the class is marked as `final`.